tf2_ros/include/tf2_ros/transform_listener.h

```diff
@@ -125,43 +126,51 @@ class TransformListener
     callback_t static_cb = std::bind(
       &TransformListener::subscription_callback, this, std::placeholders::_1, true);
 
-    message_subscription_tf_ = rclcpp::create_subscription<tf2_msgs::msg::TFMessage>(
-      node,
-      "/tf",
-      qos,
-      std::move(cb),
-      options);
-    message_subscription_tf_static_ = rclcpp::create_subscription<tf2_msgs::msg::TFMessage>(
-      node,
-      "/tf_static",
-      static_qos,
-      std::move(static_cb),
-      static_options);
-
-    if (spin_thread) {
-      initThread(node->get_node_base_interface());
+    if (spin_thread_) {
+      // Create new callback group for message_subscription of tf and tf_static
+      callback_group_ = node_base_interface_->create_callback_group(
+        rclcpp::CallbackGroupType::MutuallyExclusive, false);
+      // Duplicate to modify option of subscription
+      rclcpp::SubscriptionOptionsWithAllocator<AllocatorT> tf_options = options;
+      rclcpp::SubscriptionOptionsWithAllocator<AllocatorT> tf_static_options = static_options;
+      tf_options.callback_group = callback_group_;
+      tf_static_options.callback_group = callback_group_;
+
+      message_subscription_tf_ = rclcpp::create_subscription<tf2_msgs::msg::TFMessage>(
+        node, "/tf", qos, std::move(cb), tf_options);
+      message_subscription_tf_static_ = rclcpp::create_subscription<tf2_msgs::msg::TFMessage>(
+        node, "/tf_static", static_qos, std::move(static_cb), tf_static_options);
+
+      // Create executor with dedicated thread to spin.
+      executor_ = std::make_shared<rclcpp::executors::SingleThreadedExecutor>();
+      executor_->add_callback_group(callback_group_, node_base_interface_);
```

Comment:
I think this broke an existing use-case where we expect the node to have other callbacks processed, not just the the two subscriptions added here. E.g. previously I could write something like this:

```c++
auto my_node = std::make_shared<rclcpp::Node>("my_node");
auto sub = my_node.create_subscription<std_msgs::msg::String>("chatter", 1, [](auto & msg){
  // this callback never triggered after this PR
});

tf2_ros::TransformListener listener(buffer_core_ptr, my_node, true, qos);

while (rclcpp::ok()) {
  // Expecting the executor in TransformListener to process callbacks
  std::this_thread::sleep_for(std::chrono::seconds(3));
}
```

Maybe assuming callbacks will be handled by the transform listener is a bad assumption, but the signature where we provide our own Node (and we set `spin_thread=true`) seems ill-defined.

Reply:
`my_node` is never spun in this example -- how could the subscription receive messages? The `spin_thread` is w.r.t. the TF callbacks to isolate them from the rest of the system, its not there to process your own callbacks in your own application. That was the intent of this PR to make TF2's system decoupled from the application code. You should `spin_some()` in your while loop to resolve this issue (which you should have been doing before this PR anyway?)

Reply:
I might be wrong about what I think is happening; I will report back with a complete example.

Reply:
I've updated the code in my comment to be a compilable example. It appears that I can get the example to work if I uncomment the `rclcpp::spin_some()` call in the while-loop. IIRC, this would result in a runtime error previously. Should it be the new recommended approach?

Reply:
Yes -- before I think it was a design error to have it spin for the user application too, that seems like a breakage in separation of concerns to have TF2 processing your execution model for your application. I would have assumed that it did what it is now doing and just spin a thread for TF2 to process its information internally. This was made because we want TF2 to process its own info and not be blocked or blocking other application callbacks & reducing the number of `Node` objects in complex stacks like Nav2.

Reply:
I agree that it makes sense to not have users rely on the TransformListener executor, however the current change silently breaks anyone who was relying on this behavior (it took me a while to figure out why my callbacks were not being triggered). It would be nice if we could detect this scenario and warn users. At the very least, I think we should add a release note for ROS Humble.

Reply:
100% agree

Reply:
It does not seem straight forward to add a warning, but I've added a release note in https://github.com/ros2/ros2_documentation/pull/2127